database.go

```diff
@@ -0,0 +1,723 @@
+func (p SeriesIDs) Less(i, j int) bool { return p[i] < p[j] }
+func (p SeriesIDs) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+// Equals assumes that both are sorted. This is by design, no touchy!
+func (a SeriesIDs) Equals(seriesIDs SeriesIDs) bool {
+	if len(a) != len(seriesIDs) {
+		return false
+	}
+	for i, s := range seriesIDs {
+		if a[i] != s {
+			return false
+		}
+	}
+	return true
+}
+
+// Intersect returns a new collection of series ids in sorted order that is the intersection of the two.
+// The two collections must already be sorted.
+func (a SeriesIDs) Intersect(seriesIDs SeriesIDs) SeriesIDs {
+	l := a
+	r := seriesIDs
+
+	// we want to iterate through the shortest one and stop
+	if len(seriesIDs) < len(a) {
+		l = seriesIDs
+		r = a
+	}
+
+	// they're in sorted order so advance the counter as needed.
+	// That is, don't run comparisons against lower values that we've already passed
+	var i, j int
+
+	ids := make([]uint32, 0, len(l))
+	for i < len(l) {
+		if l[i] == r[j] {
+			ids = append(ids, l[i])
+			i += 1
+			j += 1
+		} else if l[i] < r[j] {
+			i += 1
+		} else {
+			j += 1
+		}
+	}
+
+	return SeriesIDs(ids)
+}
+
+// Union returns a new collection of series ids in sorted order that is the union of the two.
+// The two collections must already be sorted.
+func (l SeriesIDs) Union(r SeriesIDs) SeriesIDs {
+	ids := make([]uint32, 0, len(l)+len(r))
+	var i, j int
+	for i < len(l) && j < len(r) {
+		if l[i] == r[j] {
+			ids = append(ids, l[i])
+			i += 1
+			j += 1
+		} else if l[i] < r[j] {
+			ids = append(ids, l[i])
+			i += 1
+		} else {
+			ids = append(ids, r[j])
+			j += 1
+		}
+	}
+
+	// now append the remainder
+	if i < len(l) {
+		ids = append(ids, l[i:]...)
+	} else if j < len(r) {
+		ids = append(ids, r[j:]...)
+	}
+
+	return ids
+}
+
+// Reject returns a new collection of series ids in sorted order with the passed in set removed from the original. This is useful for the NOT operator.
+// The two collections must already be sorted.
+func (l SeriesIDs) Reject(r SeriesIDs) SeriesIDs {
+	var i, j int
+
+	ids := make([]uint32, 0, len(l))
+	for i < len(l) && j < len(r) {
+		if l[i] == r[j] {
+			i += 1
+			j += 1
+		} else if l[i] < r[j] {
+			ids = append(ids, l[i])
+			i += 1
+		} else {
+			j += 1
+		}
+	}
+
+	// append the remainder
+	if i < len(l) {
+		ids = append(ids, l[i:]...)
+	}
+
+	return SeriesIDs(ids)
+}
+
+// AddSeries adds the series for the given measurement to the index. Returns false if already present
+func (t *database) AddSeries(name string, s *Series) bool {
+	// if there is a measurement for this id, it's already been added
+	if t.seriesToMeasurement[s.ID] != nil {
+		return false
+	}
+
+	// get or create the measurement index and index it globally and in the measurement
+	idx := t.createMeasurementIfNotExists(name)
+
+	t.seriesToMeasurement[s.ID] = idx
+	t.series[s.ID] = s
+
+	// TODO: add this series to the global tag index
+
+	return idx.addSeries(s)
```

Comment:
In contrast to the function name above, this is perfectly clear since it's on the index object.